src/bech32.cpp

```diff
@@ -161,26 +155,28 @@ std::pair<std::string, data> Decode(const std::string& str) {
     for (size_t i = 0; i < str.size(); ++i) {
         unsigned char c = str[i];
         if (c < 33 || c > 126) return {};
-        if (c >= 'a' && c <= 'z') lower = true;
-        if (c >= 'A' && c <= 'Z') upper = true;
+        // Mixing upper and lowercase is not OK.
+        if (!lower && c >= 'a' && c <= 'z') { if(upper) return {}; lower = true; }
+        else if (!upper && c >= 'A' && c <= 'Z') { if(lower) return {}; upper = true; }
     }
-    if (lower && upper) return {};
+
     size_t pos = str.rfind('1');
     if (str.size() > 90 || pos == str.npos || pos == 0 || pos + 7 > str.size()) {
         return {};
     }
     data values(str.size() - 1 - pos);
     for (size_t i = 0; i < str.size() - 1 - pos; ++i) {
         unsigned char c = str[i + pos + 1];
-        int8_t rev = (c < 33 || c > 126) ? -1 : CHARSET_REV[c];
+        int8_t rev = CHARSET_REV[c];
```

Comment:
Add `assert(c >= 33 && c <= 126);` on the line before `int8_t rev = CHARSET_REV[c];` to make assumption explicit?

Reply:
Remember that we have asserts enabled on release notes, better not add them in inner loops especially not if the goal is to 'tighten up' anything.